client/src/state/transition.ts

```diff
@@ -0,0 +1,154 @@
+        }
+        if (homeHexalotSelected(this.appState)) {
+            const journeyPresent = homeHexalot.fetchJourney(this.appState.storage, this.appState.island, journey => {
+                // TODO: make sure that "this" is actually the latest state
+                this.subject.next(this.withJourney(journey).state)
+            })
+            console.log(`Journey present for ${homeHexalot.id}`, journeyPresent)
+        }
+        return this.withMode(Mode.Landed).withJourney(homeHexalot.journey)
+    }
+
+    public get withRestructure(): Transition {
+        const island = this.appState.island
+        island.recalculate()
+        const hexalots = island.hexalots
+        const spots = island.spots
+        const vacant = island.vacantHexalot
+        if (hexalots.length === 1) {
+            spots.forEach(spot => spot.free = true)
+        } else if (vacant) {
+            spots.forEach(spot => spot.free = spot.memberOfHexalot.every(hexalot => hexalot.id === vacant.id))
+        } else {
+            spots.forEach(spot => spot.free = false)
+        }
+        hexalots.forEach(hexalot => hexalot.refreshId())
+        const islandIsLegal = island.islandIsLegal
+        if (islandIsLegal) {
+            this.appState = {...this.appState, islandIsLegal}
+        } else {
+            this.appState = {...this.appState, islandIsLegal, mode: Mode.FixingIsland}
+        }
+        return this
+    }
+
+    public withSurface(surface: Surface): Transition {
+        const appState = this.appState
+        const selectedSpot = appState.selectedSpot
+        if (!selectedSpot) {
+            return this
+        }
+        selectedSpot.surface = surface
+        selectedSpot.memberOfHexalot.forEach(hexalot => hexalot.refreshId())
+        const nextFree = selectedSpot.adjacentSpots.find(s => s.free && s.surface === Surface.Unknown)
+        if (nextFree) {
+            return this.withSelectedSpot(nextFree)
+        }
+        const island = appState.island
+        const anyFree = island.spots.find(s => s.free && s.surface === Surface.Unknown)
+        if (anyFree) {
+            return this.withSelectedSpot(anyFree)
+        }
+        const illegal = island.spots.find(s => !s.isLegal)
+        if (illegal) {
+            return this.withSelectedSpot(illegal)
+        }
+        const vacantHexalot = island.vacantHexalot
+        if (vacantHexalot) {
+            return this.withSelectedSpot(vacantHexalot.centerSpot)
+        }
+        return this
+    }
+
+    public withGotchi(gotchi: Gotchi, journey?: Journey): Transition {
+        this.recycle()
+        const mode = journey ? Mode.DrivingJourney : Mode.DrivingFree
+        this.appState = {...this.appState, gotchi, journey, mode}
+        return this
+    }
+
+    public withEvolution(evolution: Evolution): Transition {
+        this.recycle()
+        this.appState = {...this.appState, evolution, mode: Mode.Evolving}
+        return this
+    }
+
+    private recycle(): void {
```

Comment:
What's the purpose of this?

Reply:
Prevent memory leak.  Geometries don't get collected unless disposed of properly. And also for the WASM:

`toFree => this.instanceUsed[toFree] = false`

Reply:
Got it.